### PR TITLE
Check peer-reported host against socket host

### PR DESCRIPTION
**Pull request by avieth**
This is to prevent a peer from denying service to another peer by lying
about their end point address. The attacker would give the address
of the victim and then simply hold that heavyweight connection open, so
that requests by the victim would be rejected as crossed connections.

### CODE REVIEW

File: src/Network/Transport/TCP.hs [hunk 1]
```diff
@@ -855,16 +854,33 @@ apiCloseEndPoint transport evs ourEndPoint =
...
+      <- maybe (throwIO (userError "handleConnectionRequest: peer gave malformed address"))
+               return
+               (decodeEndPointAddress theirAddress)
+    -- If the OS-determined host doesn't match the host that the peer gave us,
+    -- then we have no choice but to reject the connection. It's because we
+    -- use the EndPointAddress to key the remote end points (localConnections)
+    -- and we don't want to allow a peer to deny service to other peers by
+    -- claiming to have their host and port.
+    unless (theirHost == actualHost) $ do
+      throwIO (userError "handleConnectionRequest: reported host does not match actual host")
```
Comment (facundominguez) [inline src/Network/Transport/TCP.hs]: Please, include the values of `theirHost` and `actualHost` in the error message.

File: src/Network/Transport/TCP/Internal.hs [hunk 2]
```diff
@@ -177,7 +190,7 @@ forkServer host port backlog reuseAddr terminationHandler requestHandler = do
     acceptRequest :: N.Socket -> IO ()
     acceptRequest sock = bracketOnError (N.accept sock)
                                         (tryCloseSocket . fst)
-                                        (requestHandler . fst)
+                                        (requestHandler)
```
Comment (facundominguez) [inline src/Network/Transport/TCP/Internal.hs]: Redundant parenthesis.

File: src/Network/Transport/TCP.hs [hunk 3]
```diff
@@ -24,8 +24,6 @@ module Network.Transport.TCP
   , createTransportExposeInternals
   , TransportInternals(..)
   , EndPointId
-  , encodeEndPointAddress
-  , decodeEndPointAddress
```
Comment (facundominguez) [inline src/Network/Transport/TCP.hs]: Any reason to change the exposed interface here?

Comment (avieth) [inline src/Network/Transport/TCP.hs]: It's not necessary. Encoding/decoding an address just doesn't strike me as something that a typical user of network-transport-tcp will need to do, so I export them only from `Internal`.

File: src/Network/Transport/TCP/Internal.hs [hunk 4]
```diff
@@ -209,3 +222,47 @@ recvExact sock len = go [] len
       if BS.null bs
         then throwIO (userError "recvExact: Socket closed")
         else go (bs : acc) (l - fromIntegral (BS.length bs))
+
+-- | Produce a HostName and ServiceName from a SockAddr. Only gives 'Just' for
+-- IPv4 addresses.
+decodeSockAddr :: N.SockAddr -> IO (Maybe (N.HostName, N.ServiceName))
+decodeSockAddr sockAddr = case sockAddr of
+  N.SockAddrInet port host -> do
+    hostString <- N.inet_ntoa host
```
Comment (avieth) [inline src/Network/Transport/TCP/Internal.hs]: I'd like to allow the use of domain names. I suppose we could use `getnameinfo` to try and find the name... or, have `createTransport` resolve the name to an address when it starts up, and use that as the endpoint address.

**Comment by avieth** [comment]
Bump. I have a patch to implement #59 on top of this branch.

File: tests/TestTCP.hs [hunk 5]
```diff
@@ -694,15 +694,14 @@ testReconnect = do
...
+    wait
     case resultConnect of
       Nothing -> return ()
       Just (Left (TransportError ConnectFailed _)) -> return ()
       Just (Left err) -> throwIO err
       Just (Right _) -> throwIO $ userError "testConnect: unexpected connect success"
 
-    WaitSocketFree wait <- takeMVar socketClosed
-    wait
-
```
Comment (facundominguez) [inline tests/TestTCP.hs]: What difference does moving this before the case make?

File: tests/TestTCP.hs [hunk 6]
```diff
@@ -694,15 +694,14 @@ testReconnect = do
...
     -- Second attempt: server accepts but then closes the socket. We expect
     -- either a failed connection, or a subsequent send to fail.
     resultConnect <- timeout 500000 $ connect endpoint theirAddr ReliableOrdered defaultConnectHints
```
Comment (facundominguez) [inline tests/TestTCP.hs]: If the timeout occurs before the server gets the connection request, the test would block. I don't understand what this timeout call accomplishes.

File: tests/TestTCP.hs [hunk 7]
```diff
@@ -727,7 +726,7 @@ testReconnect = do
     -- But a send will fail because the server has closed the connection again
     WaitSocketFree wait <- takeMVar socketClosed
     wait
-    threadDelay 100000
+    threadDelay 500000
```
Comment (facundominguez) [inline tests/TestTCP.hs]: What does `threadDelay` accomplish here?

File: src/Network/Transport/TCP.hs [hunk 8]
```diff
@@ -910,6 +913,23 @@ handleConnectionRequest transport socketClosed sock = handle handleException $ d
...
+      -- use the EndPointAddress to key the remote end points (localConnections)
+      -- and we don't want to allow a peer to deny service to other peers by
+      -- claiming to have their host and port.
+      unless (theirHost == actualHost) $ do
+        let errorString = concat [
+                "handleConnectionRequest: reported host "
+              , show theirHost
+              , " does not match actual host "
+              , show actualHost
+              ]
```
Comment (facundominguez) [inline src/Network/Transport/TCP.hs]: I'm just realizing the error message is not visible to the user because `handleException` discards the exception.

**Comment by facundominguez** [comment]
As it stands, it seems a bit useless to have the peer send the external address if the transport can find it out.

It is like the following dialog:
- Alexander, what is your name?
- My name is Duncan.
- You are lying to me, I know you are Alexander! I'll shut you down.
- Heck! Why did you asked to begin with?

**Comment by avieth** [comment]
> As it stands, it seems a bit useless to have the peer send the external address if the transport can find it out.

Fair point, but the peer *does* need to send the port on which it listens, since it's not the same as the port of this socket. So we could have the peer deconstruct its own address and send only the port, but it was just easier to implement it as it is now: send the whole `EndPointAddress`.

File: tests/TestTCP.hs (hunk 7 above)
Comment (avieth) [inline tests/TestTCP.hs]: Oddly enough, if there's no delay, the send succeeds. I'm going to look into this test case and make a separate pull request to fix it. In the mean-time, #60 should be reviewed and merged.

**Comment by avieth** [comment]
Patch for `testReconnect` included in #60  https://github.com/haskell-distributed/network-transport-tcp/pull/60/commits/e0742c6d21dd193c1efde5904b836bb873860914

**Comment by avieth** [comment]
Rebased and pushed to fix conflicts. Why isn't CI running?

**Comment by facundominguez** [comment]
hm, I don't know why CI is not running.

> So we could have the peer deconstruct its own address and send only the port, but it was just easier to implement it as it is now: send the whole EndPointAddress.

I think the most practical is to not send the host. Otherwise, the remote peer will have to guess what its external address is to contact a given node.

**Comment by avieth** [comment]
> I think the most practical is to not send the host. Otherwise, the remote peer will have to guess what its external address is to contact a given node.

It won't have to guess, because it will reject when they don't match. Call it a sanity check: if the connecting `EndPoint` is misconfigured then all connections to honest peers will fail.

**Comment by facundominguez** [comment]
I think I failed to communicate the point:
If node A wants to connect to B, node A will have to guess which is the external address that node B expects node A to use.

Asking A to know which external address B expects of it, looks like offering little benefit for the trouble it makes. But I might be misjudging here.

And assuming A knows which external addresses other nodes expect of it, is there any way for A to control which external address is sent to each node?

**Comment by avieth** [comment]
> If node A wants to connect to B, node A will have to guess which is the external address that node B expects node A to use.
>
> Asking A to know which external address B expects of it, looks like offering little benefit for the trouble it makes. But I might be misjudging here.

This isn't a problem for nodes which really are addressable on the network. In the data centre, for instance, it's feasible to configure each transport with the proper address given the machine on which it runs.

The benefit is mainly for security when using a network-transport-tcp on a network that you don't completely control.

> And assuming A knows which external addresses other nodes expect of it, is there any way for A to control which external address is sent to each node?

A will always send the same address as determined by the address of its `Transport` and its `EndPointId`.

**Comment by facundominguez** [comment]
> A will always send the same address as determined by the address of its Transport and its EndPointId.

Ok, so we wouldn't allow a transport to listen for connections on multiple interfaces (say by using "0.0.0.0").

**Comment by avieth** [comment]
> Ok, so we wouldn't allow a transport to listen for connections on multiple interfaces (say by using "0.0.0.0").

This should be fine. The source IP is not checked against anything.

There would be problems if the system uses two or more interfaces for outgoing connections. It would have to send a different `EndPointAddress` depending upon which interface is used. But this is a problem even without this patch. It's assumed that there's just one network address for each transport, that all incoming connections come through one interface. Is there a good reason why outgoing connections would go through some other interface?

**Comment by facundominguez** [comment]
> Is there a good reason why outgoing connections would go through some other interface?

I can imagine an hypothetical user comming across this, but we can defer it until then,.

If this is a sanity check, I think it is fair to ask the transport to answer back to the peer with the reason for closing the connection when the addresses don't match. Otherwise the failure cause is totally invisible to the application in both the local and the remote node.

**Comment by avieth** [comment]
> If this is a sanity check, I think it is fair to ask the transport to answer back to the peer with the reason for closing the connection when the addresses don't match. Otherwise the failure cause is totally invisible to the application in both the local and the remote node.

I'm for this. I can also make this feature optional, with the default to not check the host.

**Comment by facundominguez** [comment]
Sounds good to me.

File: src/Network/Transport/TCP.hs [hunk 9]
```diff
@@ -913,43 +920,38 @@ handleConnectionRequest transport socketClosed (sock, sockAddr) = handle handleE
...
+               (decodeEndPointAddress theirAddress)
+    let checkPeerHost = tcpCheckPeerHost (transportParams transport)
+    if checkPeerHost && (theirHost /= actualHost)
+    then
+      -- If the OS-determined host doesn't match the host that the peer gave us,
+      -- then we have no choice but to reject the connection. It's because we
+      -- use the EndPointAddress to key the remote end points (localConnections)
+      -- and we don't want to allow a peer to deny service to other peers by
+      -- claiming to have their host and port.
+      sendMany sock [encodeWord32 (encodeConnectionRequestResponse ConnectionRequestHostMismatch)]
```
Comment (facundominguez) [inline src/Network/Transport/TCP.hs]: Can we have the `actualHost` included in the response to aid diagnosis?

**Comment by avieth** [comment]
Parallel connecting tests intermittently failed but I believe this has fixed it https://github.com/haskell-distributed/network-transport-tcp/pull/54/commits/6757ce91eeb3fe67cde33f1f2178e090f8348345

The bug was introduced here https://github.com/haskell-distributed/network-transport-tcp/pull/56/commits/683d3bed497b967724cfb4206fd53faa78e5bf22 and manifested as a RELY  violation in `createConnectionTo`; the remote was in a closed/closing state because there was a connection coming up which had not yet sent `CreatedNewConnection`.

**Comment by facundominguez** [comment]
Thanks @avieth!
